### PR TITLE
throw exception in item enahancement, daily reward

### PR DESCRIPTION
**Pull request by ipdae**
- reslove #131

### CODE REVIEW

**Review by boscohyun — APPROVED** [review]

File: Lib9c/Action/ItemEnhancement.cs [hunk 1]
```diff
@@ -183,83 +175,60 @@ public override IAccountStateDelta Execute(IActionContext context)
             {
                 if (!avatarState.inventory.TryGetNonFungibleItem(materialId, out ItemUsable materialItem))
                 {
-                    // 인벤토리에 재료로 등록한 장비가 없는 에러.
-                    return LogError(
-                        context,
-                        "Aborted as the the signer does not have a necessary material ({MaterialId}).",
-                        materialId
+                    throw new NotEnoughMaterialException(
+                        $"Aborted as the the signer does not have a necessary material ({materialId})."
```
Comment (moreal) [inline Lib9c/Action/ItemEnhancement.cs:179]: 👀 

```suggestion
                        $"Aborted as the signer does not have a necessary material ({materialId})."
```

File: Lib9c/Action/ItemEnhancement.cs [hunk 2]
```diff
@@ -183,83 +175,60 @@ public override IAccountStateDelta Execute(IActionContext context)
...
                 }
 
                 if (!(materialItem is Equipment materialEquipment))
                 {
-                    return LogError(
-                        context,
-                        $"Aborted as the material item is not a {nameof(Equipment)}, but {{ItemType}}.",
-                        materialItem.GetType().Name
+                    throw new InvalidCastException(
+                        $"Aborted as the material item is not a {nameof(Equipment)}, but {materialItem.GetType().Name}."
```
Comment (moreal) [inline Lib9c/Action/ItemEnhancement.cs:193]:
```suggestion
                        $"Aborted as the material item is not an {nameof(Equipment)}, but {materialItem.GetType().Name}."
```

File: Lib9c/Action/ItemEnhancement.cs [hunk 3]
```diff
@@ -183,83 +175,60 @@ public override IAccountStateDelta Execute(IActionContext context)
...
                 {
-                    // 등급이 다른 에러.
-                    return LogError(
-                        context,
-                        "Aborted as grades of the equipment to enhance ({EquipmentGrade}) and a material ({MaterialGrade}) do not match.",
-                        enhancementEquipment.Grade,
-                        materialEquipment.Grade
+                    // Invalid Grade
+                    throw new InvalidMaterialException(
+                        $"Aborted as grades of the equipment to enhance ({enhancementEquipment.Grade}) and a material ({materialEquipment.Grade}) do not match."
```
Comment (moreal) [inline Lib9c/Action/ItemEnhancement.cs:223]:
```suggestion
                        $"Aborted as grades of the equipment to enhance ({enhancementEquipment.Grade}) and a material ({materialEquipment.Grade}) does not match."
```

File: Lib9c/Action/ItemEnhancement.cs [hunk 4]
```diff
@@ -183,83 +175,60 @@ public override IAccountStateDelta Execute(IActionContext context)
...
                 {
-                    // 강화도가 다른 에러.
-                    return LogError(
-                        context,
-                        "Aborted as levels of the equipment to enhance ({EquipmentLevel}) and a material ({MaterialLevel}) do not match.",
-                        enhancementEquipment.level,
-                        materialEquipment.level
+                    // Invalid level
+                    throw new InvalidMaterialException(
+                        $"Aborted as levels of the equipment to enhance ({enhancementEquipment.level}l}}) and a material ({materialEquipment.level}) do not match."
```
Comment (moreal) [inline Lib9c/Action/ItemEnhancement.cs:231]:
```suggestion
                        $"Aborted as levels of the equipment to enhance ({enhancementEquipment.level}l}}) and a material ({materialEquipment.level}) does not match."
```

Comment (longfin) [reply Lib9c/Action/ItemEnhancement.cs:231]: `({enhancementEquipment.level}l}})` seems typo too.

**Review by Unengine — APPROVED** [review]

**Review by longfin — APPROVED** [review]

File: Lib9c/Action/ItemEnhancement.cs [hunk 5]
```diff
@@ -155,11 +150,8 @@ public override IAccountStateDelta Execute(IActionContext context)
             if (avatarState.actionPoint < requiredAP)
             {
                 // AP 부족 에러.
```
Comment (limebell) [inline Lib9c/Action/ItemEnhancement.cs:152]:
```suggestion
```

**Pull request merged**